services/galley/src/Galley/API/Clients.hs

```diff
@@ -61,9 +85,71 @@ addClientH (usr ::: clt) = do
   pure empty
 
 rmClientH ::
-  Member ClientStore r =>
+  forall p1 r.
+  ( p1 ~ CassandraPaging,
+    Members
+      '[ ClientStore,
+         ConversationStore,
+         ExternalAccess,
+         FederatorAccess,
+         GundeckAccess,
+         Input Env,
+         Input (Local ()),
+         Input UTCTime,
+         ListItems p1 ConvId,
+         ListItems p1 (Remote ConvId),
+         MemberStore,
+         Poly.Error InternalError,
+         ProposalStore,
+         P.TinyLog
+       ]
+      r
+  ) =>
   UserId ::: ClientId ->
   Sem r Response
-rmClientH (usr ::: clt) = do
-  E.deleteClient usr clt
+rmClientH (usr ::: cid) = do
+  lusr <- qualifyLocal usr
+  let nRange1000 = toRange (Proxy @1000) :: Range 1 1000 Int32
+  firstConvIds <- Query.conversationIdsPageFrom lusr (GetPaginatedConversationIds Nothing nRange1000)
+  goConvs nRange1000 firstConvIds lusr
+
+  E.deleteClient usr cid
   pure empty
+  where
+    rpc = fedClient @'Galley @"on-client-removed"
+    goConvs :: Range 1 1000 Int32 -> ConvIdsPage -> Local UserId -> Sem r ()
+    goConvs range page lusr = do
+      loc :: Local () <- input
+      let (localConvs, remoteConvs) = partitionQualified loc (mtpResults page)
+      for_ localConvs $ \convId -> do
+        mConv <- getConversation convId
+        for_ mConv $ \conv -> do
+          lconv <- qualifyLocal conv
+          removeClient lconv (qUntagged lusr) cid
+      traverse_ removeRemoteMLSClients (rangedChunks remoteConvs)
+      when (mtpHasMore page) $ do
+        let nextState = mtpPagingState page
+            nextQuery = GetPaginatedConversationIds (Just nextState) range
+        newCids <- Query.conversationIdsPageFrom lusr nextQuery
+        goConvs range newCids lusr
+
+    removeRemoteMLSClients :: Range 1 1000 [Remote ConvId] -> Sem r ()
+    removeRemoteMLSClients cids = do
+      for_ (bucketRemote (fromRange cids)) $ \remoteConvs -> do
+        runFederatedEither remoteConvs (rpc (ClientRemovedRequest usr cid (tUnqualified remoteConvs)))
+          >>= logAndIgnoreError "Error in onConversationUpdated call" usr
```

Comment:
Yes, it is. `rpc` is bound to `fedClient @'Galley @"on-client-removed"`. The reason we define this name outside is because creating the servant client `rpc` is costly apparently

Reply:
I am not sure I follow. I am asking if instead of:
```haskell
  logAndIgnoreError "Error in onConversationUpdated call" usr
```
it should have been:
```haskell
  logAndIgnoreError "Error in onClientRemoved call" usr
```
instead.